# Python Study Notebook: Asn
### Questions, Unchanged User Code, Detailed Logic & Step-by-Step Code Tracing

---

## Section 1: wap in dispaly the asn number with customized function

### ❓ Question / Problem Statement
**Problem Statement (wap in dispaly the asn number with customized function): Implement and trace the Python program for wap in dispaly the asn number with customized function.**

--- 
### 💻 My Code

In [ ]:
# wap in dispaly the asn number with customized function
def CountDigit(n):
    count=0
    while n>0:
        n=n//10
        count+=1
    return count

def isASN(n):
    temp=n
    if n<0:
        n=n*-1

    asn, pow=0, CountDigit(n)

    while n>0:
        base =n%10
        asn=asn+(base**pow)
        n=n//10

    if temp<0:
        asn=asn*-1

    return temp==asn


num=int(input("Enter the number :"))
if isASN(num):
    print(num,"is a ASN")
else:
    print(num,"is not a ASN")

153 is a ASN


### 🧠 Logic & Explanation
1. **Modular Function Design:** Defines custom function(s) `CountDigit, isASN` to encapsulate reusable task logic.
2. **User Input Handling:** Prompts user input and casts values (`int()`, `float()`) to appropriate numeric types.
3. **Modulo Division (`%`):** Uses remainder arithmetic to check divisibility, factors, or extract digits.
4. **Conditional Branching (`if-else`):** Evaluates boolean expressions to branch execution path.
5. **While Loop Control:** Continues execution while state control condition evaluates to `True`.
6. **Return Statement:** Passes calculated result back to caller function.

### 🔍 Step-by-Step Code Tracing (Dry Run)

#### Sample Execution: `n = 153`

**Step 1: Count Digits (`CountDigit(153)`):**
- `153 // 10 = 15` (1)
- `15 // 10 = 1` (2)
- `1 // 10 = 0` (3)
- `count = 3`

**Step 2: Sum of Power of Digits (`isArmstrong(153)`):**
- Digit 3: `3**3 = 27`
- Digit 5: `5**3 = 125`
- Digit 1: `1**3 = 1`
- `sum = 27 + 125 + 1 = 153`

**Step 3: Comparison (`sum == original`):**
- `153 == 153` is **True**.

**Output:**
`153 is an Armstrong number`

---

#### 💡 The Main Idea / Intuition to Remember

```
153 has 3 digits:
1³ + 5³ + 3³ = 1 + 125 + 27 = 153
Sum == Original → ARMSTRONG NUMBER!
```

---

## Section 2: wap to display all the asn's present in a user defined range

### ❓ Question / Problem Statement
**Problem Statement (wap to display all the asn's present in a user defined range): Implement and trace the Python program for wap to display all the asn's present in a user defined range.**

--- 
### 💻 My Code

In [ ]:
# wap to display all the asn's present in a user defined range 


def CountDigit(n):
    count=0
    while n>0:
        n=n//10
        count+=1
    return count

def isASN(n):
    temp=n
    if n<0:
        n=n*-1

    asn, pow=0, CountDigit(n)

    while n>0:
        base =n%10
        asn=asn+(base**pow)
        n=n//10

    if temp<0:
        asn=asn*-1

    return temp==asn

sn=int(input("enter the start number "))
en=int(input("enter the end number "))

if sn>en:
    print("invalid number")
else:
    print("asn numnber :" )
    for i in range(sn, en+1):
        if isASN(i):
            print(i, end=" ")

        



asn numnber :
153 

### 🧠 Logic & Explanation
1. **Modular Function Design:** Defines custom function(s) `CountDigit, isASN` to encapsulate reusable task logic.
2. **User Input Handling:** Prompts user input and casts values (`int()`, `float()`) to appropriate numeric types.
3. **Modulo Division (`%`):** Uses remainder arithmetic to check divisibility, factors, or extract digits.
4. **Conditional Branching (`if-else`):** Evaluates boolean expressions to branch execution path.
5. **For Loop Iteration:** Iterates through sequence/range bounds (`range(start, stop)`).
6. **While Loop Control:** Continues execution while state control condition evaluates to `True`.
7. **Return Statement:** Passes calculated result back to caller function.

### 🔍 Step-by-Step Code Tracing (Dry Run)

#### Sample Execution: `sn = 4, en = 8`

**Initially:**
- `sn = 4`, `en = 8`

**Validation (`if sn > en:`):**
- `4 > 8` is **False**, so move to `else:` branch.

**Loop (`for i in range(4, 9)`):**
- Produces sequence: `4, 5, 6, 7, 8`

- **For `i = 4`:** `isEvenOdd(4)` $\rightarrow$ `4 % 2 == 0` is **True**. Print `4 `.
- **For `i = 5`:** `isEvenOdd(5)` $\rightarrow$ `5 % 2 == 0` is **False**. Skip.
- **For `i = 6`:** `isEvenOdd(6)` $\rightarrow$ `6 % 2 == 0` is **True**. Print `6 `.
- **For `i = 7`:** `isEvenOdd(7)` $\rightarrow$ `7 % 2 == 0` is **False**. Skip.
- **For `i = 8`:** `isEvenOdd(8)` $\rightarrow$ `8 % 2 == 0` is **True**. Print `8 `.

**Output:**
`even numbers:`
`4 6 8 `

---

#### 💡 The Main Idea / Intuition to Remember

```
Range [4 to 8]:
  4 → Even? YES → Print 4
  5 → Even? NO  → Skip
  6 → Even? YES → Print 6
  7 → Even? NO  → Skip
  8 → Even? YES → Print 8
```

---

## Section 3: wap to display all the asn's present in a user defined range separalety

### ❓ Question / Problem Statement
**Problem Statement (wap to display all the asn's present in a user defined range separalety): Implement and trace the Python program for wap to display all the asn's present in a user defined range separalety.**

--- 
### 💻 My Code

In [ ]:

# wap to display all the asn's present in a user defined range separalety

def CountDigit(n):
    count=0
    while n>0:
        n=n//10
        count+=1
    return count

def isASN(n):
    temp=n
    if n<0:
        n=n*-1
    asn, pow=0, CountDigit(n)

    while n>0:
        base =n%10
        asn=asn+(base**pow)
        n=n//10

    if temp<0:
        asn=asn*-1

    return temp==asn

sn=int(input("enter the start number "))
en=int(input("enter the end number "))

if sn>en:
    print("invalid number")
else:
    print("asn numnber :" )
    for i in range(sn, en+1):
        if isASN(i):
            print(i, end=" ")


    print("\n")

    print("not a asn : ")
    for i in range (sn, en+1):
        if not isASN(i):
            print(i,end=" ")


asn numnber :
153 

not a asn : 
100 101 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144 145 146 147 148 149 150 151 152 154 155 156 157 158 159 160 161 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232 233 234 235 236 237 238 239 240 241 242 243 244 245 246 247 248 249 250 251 252 253 

### 🧠 Logic & Explanation
1. **Modular Function Design:** Defines custom function(s) `CountDigit, isASN` to encapsulate reusable task logic.
2. **User Input Handling:** Prompts user input and casts values (`int()`, `float()`) to appropriate numeric types.
3. **Modulo Division (`%`):** Uses remainder arithmetic to check divisibility, factors, or extract digits.
4. **Conditional Branching (`if-else`):** Evaluates boolean expressions to branch execution path.
5. **For Loop Iteration:** Iterates through sequence/range bounds (`range(start, stop)`).
6. **While Loop Control:** Continues execution while state control condition evaluates to `True`.
7. **Return Statement:** Passes calculated result back to caller function.

### 🔍 Step-by-Step Code Tracing (Dry Run)

#### Sample Execution: `n = 153`

**Step 1: Count Digits (`CountDigit(153)`):**
- `153 // 10 = 15` (1)
- `15 // 10 = 1` (2)
- `1 // 10 = 0` (3)
- `count = 3`

**Step 2: Sum of Power of Digits (`isArmstrong(153)`):**
- Digit 3: `3**3 = 27`
- Digit 5: `5**3 = 125`
- Digit 1: `1**3 = 1`
- `sum = 27 + 125 + 1 = 153`

**Step 3: Comparison (`sum == original`):**
- `153 == 153` is **True**.

**Output:**
`153 is an Armstrong number`

---

#### 💡 The Main Idea / Intuition to Remember

```
153 has 3 digits:
1³ + 5³ + 3³ = 1 + 125 + 27 = 153
Sum == Original → ARMSTRONG NUMBER!
```

---

## Section 4: wap to display first n asn's

### ❓ Question / Problem Statement
**Problem Statement (wap to display first n asn's): Implement and trace the Python program for wap to display first n asn's.**

--- 
### 💻 My Code

In [ ]:
# wap to display first n asn's 

def CountDigit(n):
    count=0
    while n>0:
        n=n//10
        count+=1
    return count

def isASN(n):
    temp=n
    if n<0:
        n=n*-1
    asn, pow=0, CountDigit(n)

    while n>0:
        base =n%10
        asn=asn+(base**pow)
        n=n//10

    if temp<0:
        asn=asn*-1

    return temp==asn
  


count = int(input("Enter how many Armstrong numbers: "))

series = 1

while count > 0:
    if isAsn(series):
        print(series, end=" ")
        count -= 1
    series += 1

1 2 3 4 5 6 7 8 9 153 370 371 407 1634 8208 9474 54748 92727 93084 548834 

### 🧠 Logic & Explanation
1. **Modular Function Design:** Defines custom function(s) `CountDigit, isASN` to encapsulate reusable task logic.
2. **User Input Handling:** Prompts user input and casts values (`int()`, `float()`) to appropriate numeric types.
3. **Modulo Division (`%`):** Uses remainder arithmetic to check divisibility, factors, or extract digits.
4. **Conditional Branching (`if-else`):** Evaluates boolean expressions to branch execution path.
5. **While Loop Control:** Continues execution while state control condition evaluates to `True`.
6. **Return Statement:** Passes calculated result back to caller function.

### 🔍 Step-by-Step Code Tracing (Dry Run)

#### Sample Execution: `n = 153`

**Step 1: Count Digits (`CountDigit(153)`):**
- `153 // 10 = 15` (1)
- `15 // 10 = 1` (2)
- `1 // 10 = 0` (3)
- `count = 3`

**Step 2: Sum of Power of Digits (`isArmstrong(153)`):**
- Digit 3: `3**3 = 27`
- Digit 5: `5**3 = 125`
- Digit 1: `1**3 = 1`
- `sum = 27 + 125 + 1 = 153`

**Step 3: Comparison (`sum == original`):**
- `153 == 153` is **True**.

**Output:**
`153 is an Armstrong number`

---

#### 💡 The Main Idea / Intuition to Remember

```
153 has 3 digits:
1³ + 5³ + 3³ = 1 + 125 + 27 = 153
Sum == Original → ARMSTRONG NUMBER!
```

---

## Section 5: wap to dispaly first n non asn asn's

### ❓ Question / Problem Statement
**Problem Statement (wap to dispaly first n non asn asn's): Implement and trace the Python program for wap to dispaly first n non asn asn's.**

--- 
### 💻 My Code

In [ ]:
# wap to dispaly first n non asn asn's 


def CountDigit(n):
    count=0
    while n>0:
        n=n//10
        count+=1
    return count

def isASN(n):
    temp=n
    if n<0:
        n=n*-1
    asn, pow=0, CountDigit(n)

    while n>0:
        base =n%10
        asn=asn+(base**pow)
        n=n//10

    if temp<0:
        asn=asn*-1

    return temp==asn
  


count=int(input("enter the number "))
series=1
while count>0:
    if not isAsn(series):
        print(series, end=" ")
        count-=1
    series+=1

10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 

### 🧠 Logic & Explanation
1. **Modular Function Design:** Defines custom function(s) `CountDigit, isASN` to encapsulate reusable task logic.
2. **User Input Handling:** Prompts user input and casts values (`int()`, `float()`) to appropriate numeric types.
3. **Modulo Division (`%`):** Uses remainder arithmetic to check divisibility, factors, or extract digits.
4. **Conditional Branching (`if-else`):** Evaluates boolean expressions to branch execution path.
5. **While Loop Control:** Continues execution while state control condition evaluates to `True`.
6. **Return Statement:** Passes calculated result back to caller function.

### 🔍 Step-by-Step Code Tracing (Dry Run)

#### Sample Execution: `n = 153`

**Step 1: Count Digits (`CountDigit(153)`):**
- `153 // 10 = 15` (1)
- `15 // 10 = 1` (2)
- `1 // 10 = 0` (3)
- `count = 3`

**Step 2: Sum of Power of Digits (`isArmstrong(153)`):**
- Digit 3: `3**3 = 27`
- Digit 5: `5**3 = 125`
- Digit 1: `1**3 = 1`
- `sum = 27 + 125 + 1 = 153`

**Step 3: Comparison (`sum == original`):**
- `153 == 153` is **True**.

**Output:**
`153 is an Armstrong number`

---

#### 💡 The Main Idea / Intuition to Remember

```
153 has 3 digits:
1³ + 5³ + 3³ = 1 + 125 + 27 = 153
Sum == Original → ARMSTRONG NUMBER!
```

---

## Section 6: write a code to revese to  a number using customize function

### ❓ Question / Problem Statement
**Problem Statement (write a code to revese to  a number using customize function): Implement and trace the Python program for write a code to revese to  a number using customize function.**

--- 
### 💻 My Code

In [ ]:
# write a code to revese to  a number using customize function

def integerReversal(n):
    temp=n
    if n<0:
         n=n*-1
    rev=0

    while n>0:
        rem =n%10
        rev=(rev*10)+rem
        n=n//10

    if temp<0:
        rev=rev*-1

    return rev 

num=int(input("Enter the number :"))
res = integerReversal(num)
print("the reversal of ", num,"is :", res)



the reversal of  1563 is : 3651


### 🧠 Logic & Explanation
1. **Modular Function Design:** Defines custom function(s) `integerReversal` to encapsulate reusable task logic.
2. **User Input Handling:** Prompts user input and casts values (`int()`, `float()`) to appropriate numeric types.
3. **Modulo Division (`%`):** Uses remainder arithmetic to check divisibility, factors, or extract digits.
4. **Conditional Branching (`if-else`):** Evaluates boolean expressions to branch execution path.
5. **While Loop Control:** Continues execution while state control condition evaluates to `True`.
6. **Return Statement:** Passes calculated result back to caller function.

### 🔍 Step-by-Step Code Tracing (Dry Run)

#### Sample Execution: `n = 123`

**Initially:**
- `n = 123`, `rev = 0`

**Loop (`while n > 0`):**
- **Cycle 1:** Digit `d = 123 % 10 = 3`. `rev = 0 * 10 + 3 = 3`. `n = 12`.
- **Cycle 2:** Digit `d = 12 % 10 = 2`. `rev = 3 * 10 + 2 = 32`. `n = 1`.
- **Cycle 3:** Digit `d = 1 % 10 = 1`. `rev = 32 * 10 + 1 = 321`. `n = 0`.

**Finally:**
- `return rev` returns `321`

---

#### 💡 The Main Idea / Intuition to Remember

```
Original: 123
Extract 3 → rev = 3
Extract 2 → rev = 32
Extract 1 → rev = 321
Reversed Number = 321
```

---

## Section 7: wap to reverse a number present in a user defined range

### ❓ Question / Problem Statement
**Problem Statement (wap to reverse a number present in a user defined range): Implement and trace the Python program for wap to reverse a number present in a user defined range.**

--- 
### 💻 My Code

In [ ]:
# wap to reverse a number present in a user defined range
def integerReversal(n):
    temp=n
    if n<0:
        n=n*-1
    rev=0
    
    while n>0:
        rem=n%10
        rev=(rev*10)+rem
        n=n//10

    if temp<0:
        rev=rev*-1
    
    return rev


sr=int(input("enter the  starting range : "))
er= int(input("enter the ending range: "))

if sr>er:
    print("invalid numebr")

else:
    print(" the reverse number : ")
    for i in range (sr, er+1):
        res = integerReversal(i)
        print(res ,end=" ")

 the reverse number : 
11 21 31 41 51 61 71 81 91 2 

### 🧠 Logic & Explanation
1. **Modular Function Design:** Defines custom function(s) `integerReversal` to encapsulate reusable task logic.
2. **User Input Handling:** Prompts user input and casts values (`int()`, `float()`) to appropriate numeric types.
3. **Modulo Division (`%`):** Uses remainder arithmetic to check divisibility, factors, or extract digits.
4. **Conditional Branching (`if-else`):** Evaluates boolean expressions to branch execution path.
5. **For Loop Iteration:** Iterates through sequence/range bounds (`range(start, stop)`).
6. **While Loop Control:** Continues execution while state control condition evaluates to `True`.
7. **Return Statement:** Passes calculated result back to caller function.

### 🔍 Step-by-Step Code Tracing (Dry Run)

#### Sample Execution: `n = 123`

**Initially:**
- `n = 123`, `rev = 0`

**Loop (`while n > 0`):**
- **Cycle 1:** Digit `d = 123 % 10 = 3`. `rev = 0 * 10 + 3 = 3`. `n = 12`.
- **Cycle 2:** Digit `d = 12 % 10 = 2`. `rev = 3 * 10 + 2 = 32`. `n = 1`.
- **Cycle 3:** Digit `d = 1 % 10 = 1`. `rev = 32 * 10 + 1 = 321`. `n = 0`.

**Finally:**
- `return rev` returns `321`

---

#### 💡 The Main Idea / Intuition to Remember

```
Original: 123
Extract 3 → rev = 3
Extract 2 → rev = 32
Extract 1 → rev = 321
Reversed Number = 321
```

---

## Section 8: wap to display the first n natural num in reve order

### ❓ Question / Problem Statement
**Problem Statement (wap to display the first n natural num in reve order): Implement and trace the Python program for wap to display the first n natural num in reve order.**

--- 
### 💻 My Code

In [ ]:
# wap to display the first n natural num in reve order 

# eg: n=15
# 1 2 3 4 5 6 7 7 9 1 11 21 31 41 51 

def reverNum(n):
    temp=n
    if n<0:
        n=n*-1
    rev=0

    while n>0:
        rem=n%10
        rev=(rev*10)+rem
        n=n//10

    if temp <0:
        rev=rev*-1

    return rev

count=int(input("enter the number : "))
series=1

while count>0:
    print(reverNum(series), end=" ")
    count -= 1
    series += 1


1 2 3 4 5 6 7 8 9 1 

### 🧠 Logic & Explanation
1. **Modular Function Design:** Defines custom function(s) `reverNum` to encapsulate reusable task logic.
2. **User Input Handling:** Prompts user input and casts values (`int()`, `float()`) to appropriate numeric types.
3. **Modulo Division (`%`):** Uses remainder arithmetic to check divisibility, factors, or extract digits.
4. **Conditional Branching (`if-else`):** Evaluates boolean expressions to branch execution path.
5. **While Loop Control:** Continues execution while state control condition evaluates to `True`.
6. **Return Statement:** Passes calculated result back to caller function.

### 🔍 Step-by-Step Code Tracing (Dry Run)

#### Sample Execution Walkthrough

**Initially:**
- PVM loads variables and function definitions into memory.

**Execution:**
- Evaluates statements sequentially, executing conditions and loops.

**Output:**
- Returns calculated result or displays formatted stdout output.

---

#### 💡 The Main Idea / Intuition to Remember
- Sequential execution guarantees state updates at every step.

---

## Section 9: cheak weather the number is palindrome or not

### ❓ Question / Problem Statement
**Problem Statement (cheak weather the number is palindrome or not): Implement and trace the Python program for cheak weather the number is palindrome or not.**

--- 
### 💻 My Code

In [ ]:
# 1. cheak weather the number is palindrome or not 
def num_Palindrome(n):
    temp=n
    if n<0:
        n=n*-1
    rev=0
    
    while n>0:
        rem=n%10
        rev=(rev*10)+rem
        n=n//10

    if temp<0:
        rev=rev*-1

    return rev==temp


num=int(input("enter the numbr : "))
if num_Palindrome(num):
    print("the palindrome number is",num)

else:
    print("not a palindrome number ", num )
  

not a palindrome number  65898


### 🧠 Logic & Explanation
1. **Modular Function Design:** Defines custom function(s) `num_Palindrome` to encapsulate reusable task logic.
2. **User Input Handling:** Prompts user input and casts values (`int()`, `float()`) to appropriate numeric types.
3. **Modulo Division (`%`):** Uses remainder arithmetic to check divisibility, factors, or extract digits.
4. **Conditional Branching (`if-else`):** Evaluates boolean expressions to branch execution path.
5. **While Loop Control:** Continues execution while state control condition evaluates to `True`.
6. **Return Statement:** Passes calculated result back to caller function.

### 🔍 Step-by-Step Code Tracing (Dry Run)

#### Sample Execution: `n = 121`

**Step 1: Reverse Number:**
- `rev(121)` returns `121`.

**Step 2: Comparison (`original == reversed`):**
- `121 == 121` is **True**.

**Output:**
`121 is a Palindrome number`

---

#### 💡 The Main Idea / Intuition to Remember

```
Original Number: 121
Reversed Number: 121
Original == Reversed → PALINDROME!
```

---

## Section 10: cheak weather the number is palindrome in the user range

### ❓ Question / Problem Statement
**Problem Statement (cheak weather the number is palindrome in the user range): Implement and trace the Python program for cheak weather the number is palindrome in the user range.**

--- 
### 💻 My Code

In [ ]:
# 2. cheak weather the number is palindrome in the user range

def num_Palindrome(n):
    temp=n
    if n<0:
        n=n*-1
    rev=0

    while n>0:
        rem=n%10
        rev=(rev*10)+rem  
        n=n//10
    
    if temp <0:
        rev=rev*-1
    

    return rev==temp

sr=int(input("enter the starting number : "))
er = int(input("enter the end number : "))

if sr >er:
    print("invalided number ")

else:
    print("the first palindrome number : ")
    for i in range (sr, er+1):
       if num_Palindrome(i):
        print(i,end=" ")

    



the first palindrome number : 
101 111 121 131 141 151 161 171 181 191 202 212 222 232 242 252 262 272 282 292 

### 🧠 Logic & Explanation
1. **Modular Function Design:** Defines custom function(s) `num_Palindrome` to encapsulate reusable task logic.
2. **User Input Handling:** Prompts user input and casts values (`int()`, `float()`) to appropriate numeric types.
3. **Modulo Division (`%`):** Uses remainder arithmetic to check divisibility, factors, or extract digits.
4. **Conditional Branching (`if-else`):** Evaluates boolean expressions to branch execution path.
5. **For Loop Iteration:** Iterates through sequence/range bounds (`range(start, stop)`).
6. **While Loop Control:** Continues execution while state control condition evaluates to `True`.
7. **Return Statement:** Passes calculated result back to caller function.

### 🔍 Step-by-Step Code Tracing (Dry Run)

#### Sample Execution: `n = 121`

**Step 1: Reverse Number:**
- `rev(121)` returns `121`.

**Step 2: Comparison (`original == reversed`):**
- `121 == 121` is **True**.

**Output:**
`121 is a Palindrome number`

---

#### 💡 The Main Idea / Intuition to Remember

```
Original Number: 121
Reversed Number: 121
Original == Reversed → PALINDROME!
```

---

## Section 11: wap of the first plaindrome number

### ❓ Question / Problem Statement
**Problem Statement (wap of the first plaindrome number): Implement and trace the Python program for wap of the first plaindrome number.**

--- 
### 💻 My Code

In [ ]:
# 3.wap of the first plaindrome number 

def num_Palindrome(n):
    temp=n
    if n<0:
        n=n*-1
    rev=0

    while n>0:
        rem=n%10
        rev=(rev*10)+rem  
        n=n//10
    
    if temp <0:
        rev=rev*-1
    

    return rev==temp

count=int(input("enter the number: "))
series=1

while count>0:
    if num_Palindrome(series):
        print(series,end=" ")
        count-=1

    series+=1

1 2 3 4 5 6 7 8 9 11 22 33 44 55 66 77 88 99 101 111 

### 🧠 Logic & Explanation
1. **Modular Function Design:** Defines custom function(s) `num_Palindrome` to encapsulate reusable task logic.
2. **User Input Handling:** Prompts user input and casts values (`int()`, `float()`) to appropriate numeric types.
3. **Modulo Division (`%`):** Uses remainder arithmetic to check divisibility, factors, or extract digits.
4. **Conditional Branching (`if-else`):** Evaluates boolean expressions to branch execution path.
5. **While Loop Control:** Continues execution while state control condition evaluates to `True`.
6. **Return Statement:** Passes calculated result back to caller function.

### 🔍 Step-by-Step Code Tracing (Dry Run)

#### Sample Execution: `n = 121`

**Step 1: Reverse Number:**
- `rev(121)` returns `121`.

**Step 2: Comparison (`original == reversed`):**
- `121 == 121` is **True**.

**Output:**
`121 is a Palindrome number`

---

#### 💡 The Main Idea / Intuition to Remember

```
Original Number: 121
Reversed Number: 121
Original == Reversed → PALINDROME!
```

---

## Section 12: wap of the first  not a plaindrome number

### ❓ Question / Problem Statement
**Problem Statement (wap of the first  not a plaindrome number): Implement and trace the Python program for wap of the first  not a plaindrome number.**

--- 
### 💻 My Code

In [ ]:
 # 4.wap of the first  not a plaindrome number 

def num_Palindrome(n):
    temp=n
    if n<0:
        n=n*-1
    rev=0

    while n>0:
        rem=n%10
        rev=(rev*10)+rem  
        n=n//10
    
    if temp <0:
        rev=rev*-1
    

    return rev==temp

count=int(input("enter the number: "))
series=1

while count>0:
    if not num_Palindrome(series):
        print(series,end=" ")
        count-=1

    series+=1



10 12 13 14 15 16 17 18 19 20 21 23 24 25 26 27 28 29 30 31 

### 🧠 Logic & Explanation
1. **Modular Function Design:** Defines custom function(s) `num_Palindrome` to encapsulate reusable task logic.
2. **User Input Handling:** Prompts user input and casts values (`int()`, `float()`) to appropriate numeric types.
3. **Modulo Division (`%`):** Uses remainder arithmetic to check divisibility, factors, or extract digits.
4. **Conditional Branching (`if-else`):** Evaluates boolean expressions to branch execution path.
5. **While Loop Control:** Continues execution while state control condition evaluates to `True`.
6. **Return Statement:** Passes calculated result back to caller function.

### 🔍 Step-by-Step Code Tracing (Dry Run)

#### Sample Execution: `n = 121`

**Step 1: Reverse Number:**
- `rev(121)` returns `121`.

**Step 2: Comparison (`original == reversed`):**
- `121 == 121` is **True**.

**Output:**
`121 is a Palindrome number`

---

#### 💡 The Main Idea / Intuition to Remember

```
Original Number: 121
Reversed Number: 121
Original == Reversed → PALINDROME!
```

---

## Section 13: def rev(n):

### ❓ Question / Problem Statement
**Problem Statement (def rev(n):): Implement and trace the Python program for def rev(n):.**

--- 
### 💻 My Code

In [ ]:
def rev(n):
    temp=n
    a=0
    if n<0:
        n*=-1
    
    while(n>0):
        b=n%10
        a=(a*10)+b
        n//=10
    if temp<0:
        a*=-1
    return a
n=int(input())
q=rev(n)
print(q)

4321


### 🧠 Logic & Explanation
1. **Modular Function Design:** Defines custom function(s) `rev` to encapsulate reusable task logic.
2. **User Input Handling:** Prompts user input and casts values (`int()`, `float()`) to appropriate numeric types.
3. **Modulo Division (`%`):** Uses remainder arithmetic to check divisibility, factors, or extract digits.
4. **Conditional Branching (`if-else`):** Evaluates boolean expressions to branch execution path.
5. **Return Statement:** Passes calculated result back to caller function.

### 🔍 Step-by-Step Code Tracing (Dry Run)

#### Sample Execution: `n = 123`

**Initially:**
- `n = 123`, `rev = 0`

**Loop (`while n > 0`):**
- **Cycle 1:** Digit `d = 123 % 10 = 3`. `rev = 0 * 10 + 3 = 3`. `n = 12`.
- **Cycle 2:** Digit `d = 12 % 10 = 2`. `rev = 3 * 10 + 2 = 32`. `n = 1`.
- **Cycle 3:** Digit `d = 1 % 10 = 1`. `rev = 32 * 10 + 1 = 321`. `n = 0`.

**Finally:**
- `return rev` returns `321`

---

#### 💡 The Main Idea / Intuition to Remember

```
Original: 123
Extract 3 → rev = 3
Extract 2 → rev = 32
Extract 1 → rev = 321
Reversed Number = 321
```

---